hydroengine_service/dgds_functions.py

```diff
@@ -0,0 +1,427 @@
+
+    def radians(image):
+        return ee.Image(image).toFloat().multiply(3.1415927).divide(180)
+
+    # Compute terrain properties
+    terrain = ee.Algorithms.Terrain(z)
+    slope = radians(terrain.select(['slope']))
+    aspect = radians(terrain.select(['aspect'])).resample('bicubic')
+    azimuth = radians(ee.Image.constant(azimuth))
+    zenith = radians(ee.Image.constant(zenith))
+    # hillshade
+    hs = (
+        azimuth
+            .subtract(aspect)
+            .cos()
+            .multiply(slope.sin())
+            .multiply(zenith.sin())
+            .add(
+            zenith
+                .cos()
+                .multiply(
+                slope.cos()
+            )
+        )
+            .resample('bicubic')
+    )
+
+    # weighted average of hillshade and value
+    intensity = hs.multiply(hsv.select('value'))
+
+    hue = hsv.select('hue')
+
+    # desaturate a bit
+    sat = hsv.select('saturation').multiply(sat_multiply)
+    # make a bit darker
+    val = intensity.multiply(val_multiply)
+
+    hillshaded = ee.Image.cat(hue, sat, val).hsvToRgb()
+
+    info = {}
+    info['dataset'] = 'gebco'
+    info['band'] = band
+
+    m = hillshaded.getMapId()
+    mapid = m.get('mapid')
+    token = m.get('token')
+
+    url = 'https://earthengine.googleapis.com/map/{mapid}/{{z}}/{{x}}/{{y}}?token={token}'.format(
+        mapid=mapid,
+        token=token
+    )
+
+    linear_gradient = []
+    palette = data_params['bathy_vis_params']['palette'] + data_params['topo_vis_params']['palette']
+    n_colors = len(palette)
+    offsets = np.linspace(0, 100, num=n_colors)
+    for color, offset in zip(palette, offsets):
+        linear_gradient.append({
+            'offset': '{:.3f}%'.format(offset),
+            'opacity': 100,
+            'color': color
+        })
+
+    info.update({
+        'url': url,
+        'linearGradient': linear_gradient,
+        'min': data_params['bathy_vis_params']['min'],
+        'max': data_params['topo_vis_params']['max'],
+        'imageId': source
+    })
+    return info
+
+def _generate_image_info(im, params):
+    """"generate url and tokens for image"""
+    image = ee.Image(im)
+
+    if 'sld_style' in params:
+        m = image.sldStyle(params.get('sld_style'))
+        del params['sld_style']
+    elif 'palette' in params:
+        m = image.visualize(**{
+            'min': params.get('min'),
+            'max': params.get('max'),
+            'palette': params.get('palette')
+        })
+    else:
+        m = image
+
+    if 'hillshade' in params:
+        # also pass along hillshade arguments
+        hillshade_args = params.get('hillshade_args', {})
+        m = hillshade(m, image, False, **hillshade_args)
+
+    m = m.getMapId()
+    mapid = m.get('mapid')
+    token = m.get('token')
+
+    url = 'https://earthengine.googleapis.com/map/{mapid}/{{z}}/{{x}}/{{y}}?token={token}'.format(
+        mapid=mapid,
+        token=token
+    )
+
+    linear_gradient = []
```

Comment:
You could move this above the getMapId() and merge it with the existing `elif 'palette' at r201?